test/Validate.CFG.cpp

```diff
@@ -707,9 +707,178 @@ TEST_F(ValidateCFG, DISABLED_BackEdgeBlockDoesntPostDominateContinueTargetBad) {
+  Block be_block("be_block");
+  Block merge("merge", SpvOpReturn);
+  Block exit("exit", SpvOpReturn);
+
+  loop.setBody(
+      "%cond = OpSLessThan %intt %one %two\n"
+      "OpLoopMerge %merge %cheader None\n");
+
+  string str = header + nameOps("cheader", "be_block") + types_consts +
+               "%func    = OpFunction %voidt None %funct\n";
+
+  str += entry >> loop;
+  str += loop >> vector<Block>({cheader, merge});
+  str += cheader >> vector<Block>({exit, be_block});
+  str += exit;  //  Branches out of a continue construct
+  str += be_block >> loop;
+  str += merge;
+  str += "OpFunctionEnd";
+
+  CompileSuccessfully(str);
+  ASSERT_EQ(SPV_ERROR_INVALID_CFG, ValidateInstructions());
+  EXPECT_THAT(getDiagnosticString(),
+              MatchesRegex("The continue construct with the continue target "
+                           ".\\[cheader\\] does not post dominate the "
+                           "back-edge block .\\[be_block\\]"));
+}
+
+TEST_F(ValidateCFG, BranchOutOfConstructToMergeBad) {
+  Block entry("entry");
+  Block loop("loop", SpvOpBranchConditional);
+  Block cont("cont", SpvOpBranchConditional);
+  Block merge("merge", SpvOpReturn);
+
+  loop.setBody(
+      "%cond = OpSLessThan %intt %one %two\n"
+      "OpLoopMerge %merge %loop None\n");
+
+  string str = header + nameOps("cont", "loop") + types_consts +
+               "%func    = OpFunction %voidt None %funct\n";
+
+  str += entry >> loop;
+  str += loop >> vector<Block>({cont, merge});
+  str += cont >> vector<Block>({loop, merge});
+  str += merge;
+  str += "OpFunctionEnd";
+
+  CompileSuccessfully(str);
+  ASSERT_EQ(SPV_ERROR_INVALID_CFG, ValidateInstructions());
+  EXPECT_THAT(getDiagnosticString(),
+              MatchesRegex("The continue construct with the continue target "
+                           ".\\[loop\\] does not post dominate the "
+                           "back-edge block .\\[cont\\]"));
+}
+
+TEST_F(ValidateCFG, BranchOutOfConstructBad) {
+  Block entry("entry");
+  Block loop("loop", SpvOpBranchConditional);
+  Block cont("cont", SpvOpBranchConditional);
+  Block merge("merge");
+  Block exit("exit", SpvOpReturn);
+
+  loop.setBody(
+      "%cond = OpSLessThan %intt %one %two\n"
+      "OpLoopMerge %merge %loop None\n");
+
+  string str = header + nameOps("cont", "loop") + types_consts +
+               "%func    = OpFunction %voidt None %funct\n";
+
+  str += entry >> loop;
+  str += loop >> vector<Block>({cont, merge});
+  str += cont >> vector<Block>({loop, exit});
+  str += merge >> exit;
+  str += exit;
+  str += "OpFunctionEnd";
+
+  CompileSuccessfully(str);
+  ASSERT_EQ(SPV_ERROR_INVALID_CFG, ValidateInstructions());
+  EXPECT_THAT(getDiagnosticString(),
+              MatchesRegex("The continue construct with the continue target "
+                           ".\\[loop\\] does not post dominate the "
+                           "back-edge block .\\[cont\\]"));
 }
 
 /// TODO(umar): Switch instructions
```

Comment:
Existing tests that check structured control flow rules should all use OpCapabilityShader.
Need at least one test with unstructured code, e.g. a loop with two backedges, and no OpLoopMerge or OpSelectionMerge instructions.